src/Servers/Kestrel/Core/src/Internal/Infrastructure/AsyncEnumerableReader.cs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
+{
+    internal class AsyncEnumerableReader : IValueTaskSource<int>
+    {
+        private readonly Action _onCompletedAction;
+
+        private ManualResetValueTaskSourceCore<int> _mrvts;
+        private IAsyncEnumerable<int> _readerSource;
+        private IAsyncEnumerator<int> _reader;
+
+        public Memory<byte> Buffer { get; private set; }
+        public CancellationToken CancellationToken { get; private set; }
+
+        private ValueTaskAwaiter<bool> _readAwaiter;
+
+        private volatile bool _inProgress;
+        public bool InProgress => _inProgress;
+
+        public AsyncEnumerableReader()
+        {
+            _onCompletedAction = OnCompleted;
+        }
+
+        internal void Initialize(IAsyncEnumerable<int> readerSource)
+        {
+            _readerSource = readerSource;
+            _reader = readerSource.GetAsyncEnumerator();
+        }
+
+        public ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            if (_readerSource is null)
+            {
+                ThrowNotInitialized();
+            }
+
+            if (_inProgress)
+            {
+                ThrowConcurrentReadsNotSupported();
+            }
+            _inProgress = true;
+
+            Buffer = buffer;
+            CancellationToken = cancellationToken;
+
+            var task = _reader.MoveNextAsync();
+            _readAwaiter = task.GetAwaiter();
+
+            return new ValueTask<int>(this, _mrvts.Version);
+        }
+
+        int IValueTaskSource<int>.GetResult(short token)
+        {
+            var isValid = token == _mrvts.Version;
+            try
+            {
+                return _mrvts.GetResult(token);
+            }
+            finally
+            {
+                if (isValid)
+                {
+                    Buffer = default;
+                    CancellationToken = default;
+                    _inProgress = false;
+                    _mrvts.Reset();
+                }
+            }
+        }
+
+        ValueTaskSourceStatus IValueTaskSource<int>.GetStatus(short token)
+            => _mrvts.GetStatus(token);
+
+        void IValueTaskSource<int>.OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            if (!InProgress)
+            {
+                ThrowNoReadInProgress();
+            }
+
+            _mrvts.OnCompleted(continuation, state, token, flags);
+
+            if ((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) != 0)
```

Comment:
This isn't required, always do unsafe since _mrvts will handle save and restore.

Reply:
Changed